tests/test-json-schema-to-grammar.cpp

```diff
@@ -0,0 +1,304 @@
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <regex>
+
+// TODO: split this to some library? common would need json.hpp
+#include "../examples/server/json-schema-to-grammar.h"
+#include "../examples/server/json-schema-to-grammar.cpp"
+
+using namespace std;
+
+string INPUT_NAME(tmpnam(nullptr));
+string OUT_NAME(tmpnam(nullptr));
+
+static std::string trim(const std::string & source) {
+    std::string s(source);
+    s.erase(0,s.find_first_not_of(" \n\r\t"));
+    s.erase(s.find_last_not_of(" \n\r\t")+1);
+    return regex_replace(s, regex("(^|\n)[ \t]+"), "$1");
+}
+
+struct TestCase {
+  string name;
+  string schema;
+  string expected;
+
+  void prepare() const {
+    ofstream f;
+    f.open(INPUT_NAME);
+    f << schema.c_str();
+    f.close();
+  }
+
+  void read_and_verify(const string& series) const {
+    ostringstream actuals;
+    actuals << ifstream(OUT_NAME).rdbuf();
+    auto actual = actuals.str();
+    verify(series, actual);
+  }
+
+  void verify(const string& series, const string& actual) const {
+    if (trim(actual) != trim(expected)) {
+      cerr << "#" << endl;
+      cerr << "# Test " << name.c_str() << " (" << series.c_str() << ") failed." << endl;
+      cerr << "#" << endl;
+      cerr << schema.c_str() << endl;
+      cerr << "# EXPECTED:\n" << expected.c_str() << endl;
+      cerr << "# ACTUAL:\n" << actual.c_str() << endl;
+      assert(false);
+    }
+  }
+
+};
+
+
+static void run_py(const TestCase& tc) {
+  cerr << "# Running Python " << tc.name.c_str() << endl;
+  tc.prepare();
+  assert(std::system(("python ./examples/json-schema-to-grammar.py " + INPUT_NAME + " > " + OUT_NAME).c_str()) == 0);
+  tc.read_and_verify("Python");
+}
+
+static void run_mjs(const TestCase& tc) {
+  cerr << "# Running MJS " << tc.name.c_str() << endl;
+  tc.prepare();
+  assert(std::system(("node ./tests/run-json-schema-to-grammar.mjs " + INPUT_NAME + " > " + OUT_NAME).c_str()) == 0);
+  tc.read_and_verify("JavaScript");
+}
```

Comment:
Doing this kind of string manipulation with the input to std::system is a really bad idea.

Reply:
Good point, looks smelly even in tests, now using `.tmp` files and static strings.